File: easypermissions/src/main/java/za/co/bwmuller/easypermissions/PermissionsHelper.java
package za.co.bwmuller.easypermissions;

import android.Manifest;
import android.app.Activity;
import android.app.Fragment;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import androidx.annotation.NonNull;
import androidx.annotation.StringDef;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.ArrayList;
import java.util.Arrays;

public class PermissionsHelper extends PermissionListener {
    public static final String K_READ_CONTACTS = Manifest.permission.READ_CONTACTS;
    public static final String K_RECORD_AUDIO = Manifest.permission.RECORD_AUDIO;
    public static final String K_ACCESS_FINE_LOCATION = Manifest.permission.ACCESS_FINE_LOCATION;
    public static final String K_ACCESS_COARSE_LOCATION = Manifest.permission.ACCESS_COARSE_LOCATION;
    public static final String K_INTERNET = Manifest.permission.INTERNET;
    public static final String K_ACCESS_NETWORK_STATE = Manifest.permission.ACCESS_NETWORK_STATE;
    public static final String K_VIBRATE = Manifest.permission.VIBRATE;
    public static final String K_GET_ACCOUNTS = Manifest.permission.GET_ACCOUNTS;
    public static final String K_WAKE_LOCK = Manifest.permission.WAKE_LOCK;
    public static final String K_MODIFY_AUDIO_SETTINGS = Manifest.permission.MODIFY_AUDIO_SETTINGS;
    public static final String K_READ_PHONE_STATE = Manifest.permission.READ_PHONE_STATE;
    public static final String K_WRITE_CONTACTS = Manifest.permission.WRITE_CONTACTS;
    public static final String K_READ_EXTERNAL_STORAGE = Manifest.permission.READ_EXTERNAL_STORAGE;
    public static final String K_WRITE_EXTERNAL_STORAGE = Manifest.permission.WRITE_EXTERNAL_STORAGE;
    public static final String K_CAMERA = Manifest.permission.CAMERA;
    public static final String K_READ_SMS = Manifest.permission.READ_SMS;
    private static ArrayList<PermissionRequest> permissionRequests = new ArrayList<>();

    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        PermissionRequest requestResult = new PermissionRequest(requestCode);
        if (permissionRequests.contains(requestResult)) {
            PermissionRequest permissionRequest = permissionRequests.get(permissionRequests.indexOf(requestResult));
            if (verifyPermissions(grantResults)) {
                //Permission has been granted
                permissionRequest.getPermissionCallback().permissionGranted();
            } else {
                permissionRequest.getPermissionCallback().permissionRefused();
            }
            permissionRequests.remove(requestResult);
        }
    }

    @Override public Activity getActivity() {
        return null;
    }

    private static boolean checkPermission(Context context, String permission) {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && ContextCompat.checkSelfPermission(context, permission) != PackageManager.PERMISSION_GRANTED;
    }

    /**
     * Returns false if the Context does not have access to a all given permission.
     */
    public static boolean needPermission(Context context, @PermissionsHelper.Permission String... permissions) {
        for (String permission : permissions) {
            if (checkPermission(context, permission)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Returns true if the Context has access to a all given permission.
     */
    public static boolean hasPermission(Context context, @PermissionsHelper.Permission String... permissions) {
        return !needPermission(context, permissions);
    }

    public static void requestPermissions(AppCompatCallbackActivity activity, @PermissionsHelper.Permission String[] permissions, PermissionCallback callback) {
        PermissionsHelper permissionsHelper = new PermissionsHelper();
        activity.addListener(permissionsHelper);
        permissionsHelper.askForPermission(activity, permissions, callback);
    }

    public static void requestPermission(AppCompatCallbackActivity activity, @PermissionsHelper.Permission String permissions, PermissionCallback callback) {
        PermissionsHelper permissionsHelper = new PermissionsHelper();
        activity.addListener(permissionsHelper);
        permissionsHelper.askForPermission(activity, permissions, callback);
    }

    public void askForPermission(Activity activity, @PermissionsHelper.Permission String permission) {
        askForPermission(activity, new String[]{permission}, new PermissionCallback() {
            @Override
            public void permissionGranted() {

            }

            @Override
            public void permissionRefused() {

            }
        });
    }

    public void askForPermission(Activity activity, @PermissionsHelper.Permission String[] permissions) {
        askForPermission(activity, permissions, new PermissionCallback() {
            @Override
            public void permissionGranted() {

            }

            @Override
            public void permissionRefused() {

            }
        });
    }

    public void askForPermission(Activity activity, @PermissionsHelper.Permission String permission, PermissionCallback permissionCallback) {
        askForPermission(activity, new String[]{permission}, permissionCallback);
    }

    public void askForPermission(Activity activity, @PermissionsHelper.Permission String[] permissions, PermissionCallback permissionCallback) {
        if (permissionCallback == null) {
            return;
        }
        if (!needPermission(activity, permissions)) {
            permissionCallback.permissionGranted();
            return;
        }
        PermissionRequest permissionRequest = new PermissionRequest(new ArrayList<>(Arrays.asList(permissions)), permissionCallback);
        permissionRequests.add(permissionRequest);

        requestPermissions(activity,
                permissions,
                permissionRequest.getRequestCode());
    }

    public void askForPermission(Fragment fragment, @PermissionsHelper.Permission String permission) {
        askForPermission(fragment.getActivity(), permission);
    }

    public void askForPermission(Fragment fragment, @PermissionsHelper.Permission String permission, PermissionCallback permissionCallback) {
        askForPermission(fragment.getActivity(), new String[]{permission}, permissionCallback);
    }

    private String[] getFilteredPermissions(Activity activity, @PermissionsHelper.Permission String[] permissions) {
        ArrayList<String> permissionList = new ArrayList<>();
        for (String permission : permissions) {
            if (PermissionsHelper.needPermission(activity, permission))
                permissionList.add(permission);

        }
        return permissionList.toArray(new String[permissionList.size()]);
    }

    private int[] generateFailedGrant(Activity activity, @PermissionsHelper.Permission String[] permissions) {
        int[] results = new int[permissions.length];
        for (int i = 0; i < permissions.length; i++) {
            if (PermissionsHelper.needPermission(activity, permissions[i]))
                results[i] = PackageManager.PERMISSION_DENIED;
            else
                results[i] = PackageManager.PERMISSION_GRANTED;
        }
        return results;
    }

    private void requestPermissions(Activity activity, @PermissionsHelper.Permission String[] permissionsRequired, int requestCode) {
        String[] permissions = getFilteredPermissions(activity, permissionsRequired);
        if (permissions.length > 0)
            ActivityCompat.requestPermissions(activity, permissions, requestCode);
        else
            onRequestPermissionsResult(requestCode, permissionsRequired, generateFailedGrant(activity, permissionsRequired));
    }

    /**
     * Check that all given permissions have been granted by verifying that each entry in the
     * given array is of the value {@link PackageManager#PERMISSION_GRANTED}.
     */
    public boolean verifyPermissions(int[] grantResults) {
        for (int result : grantResults) {
            if (result != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }
        return true;
    }

    public void askPermission(Activity activity, int requestCode, @PermissionsHelper.Permission String permission) {
        if (needPermission(activity, permission))
            requestPermissions(activity,
                    new String[]{permission},
                    requestCode);
    }

    /**
     * @hide
     */
    @StringDef({K_READ_CONTACTS, K_RECORD_AUDIO, K_ACCESS_FINE_LOCATION
            , K_INTERNET, K_ACCESS_NETWORK_STATE, K_VIBRATE
            , K_GET_ACCOUNTS, K_WAKE_LOCK, K_MODIFY_AUDIO_SETTINGS
            , K_READ_PHONE_STATE, K_WRITE_CONTACTS, K_READ_EXTERNAL_STORAGE
            , K_WRITE_EXTERNAL_STORAGE, K_CAMERA, K_READ_SMS
            , K_ACCESS_COARSE_LOCATION})
    @Retention(RetentionPolicy.SOURCE)
    public @interface Permission {
    }
}

File: easypermissions/src/main/java/za/co/bwmuller/easypermissions/PermissionsHelper.kt
package za.co.bwmuller.easypermissions

import android.Manifest.permission
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import androidx.annotation.StringDef
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import java.util.ArrayList

class PermissionsHelper : PermissionListener() {

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String?>, grantResults: IntArray) {
        val requestResult = PermissionRequest(requestCode)
        if (permissionRequests.contains(requestResult)) {
            val permissionRequest = permissionRequests[permissionRequests.indexOf(requestResult)]
            if (verifyPermissions(grantResults)) {
                //Permission has been granted
                permissionRequest.permissionCallback?.permissionGranted()
            } else {
                permissionRequest.permissionCallback?.permissionRefused()
            }
            permissionRequests.remove(requestResult)
        }
    }

    override val activity: Activity?
        get() = null

    fun askForPermission(
        activity: Activity, permission: String, permissionCallback: PermissionCallback = object : PermissionCallback {
            override fun permissionGranted() {}
            override fun permissionRefused() {}
        }
    ) {
        askForPermission(activity, arrayOf(permission), permissionCallback)
    }

    fun askForPermission(
        activity: Activity, permissions: Array<String>, permissionCallback: PermissionCallback = object : PermissionCallback {
            override fun permissionGranted() {}
            override fun permissionRefused() {}
        }
    ) {
        if (!needPermission(activity, *permissions)) {
            permissionCallback.permissionGranted()
            return
        }
        val permissionRequest = PermissionRequest(permissions.toMutableList(), permissionCallback)
        permissionRequests.add(permissionRequest)
        requestPermissions(
            activity,
            permissions,
            permissionRequest.requestCode
        )
    }

    fun askForPermission(
        fragment: Fragment, permission: String, permissionCallback: PermissionCallback = object : PermissionCallback {
            override fun permissionGranted() {}
            override fun permissionRefused() {}
        }
    ) {
        askForPermission(fragment.requireActivity(), arrayOf(permission), permissionCallback)
    }

    private fun getFilteredPermissions(activity: Activity, permissions: Array<String>): Array<String> {
        return permissions.filter { needPermission(activity, it) }.toTypedArray()
    }

    private fun generateFailedGrant(activity: Activity, permissions: Array<String>): IntArray {
        val results = IntArray(permissions.size)
        for (i in permissions.indices) {
            if (needPermission(activity, permissions[i])) results[i] = PackageManager.PERMISSION_DENIED else results[i] = PackageManager.PERMISSION_GRANTED
        }
        return results
    }

    private fun requestPermissions(activity: Activity, permissionsRequired: Array<String>, requestCode: Int) {
        val permissions = getFilteredPermissions(activity, permissionsRequired)
        if (permissions.isNotEmpty()) ActivityCompat.requestPermissions(activity, permissions, requestCode) else onRequestPermissionsResult(
            requestCode,
            permissionsRequired.mapTo(ArrayList<String?>()) { it }.toTypedArray(),
            generateFailedGrant(activity, permissionsRequired)
        )
    }

    /**
     * Check that all given permissions have been granted by verifying that each entry in the
     * given array is of the value [PackageManager.PERMISSION_GRANTED].
     */
    fun verifyPermissions(grantResults: IntArray): Boolean {
        for (result in grantResults) {
            if (result != PackageManager.PERMISSION_GRANTED) {
                return false
            }
        }
        return true
    }

    fun askPermission(activity: Activity, requestCode: Int, permission: String) {
        if (needPermission(activity, permission)) requestPermissions(
            activity, arrayOf(permission),
            requestCode
        )
    }

    /**
     * @hide
     */
    @StringDef(
        K_READ_CONTACTS,
        K_RECORD_AUDIO,
        K_ACCESS_FINE_LOCATION,
        K_INTERNET,
        K_ACCESS_NETWORK_STATE,
        K_VIBRATE,
        K_GET_ACCOUNTS,
        K_WAKE_LOCK,
        K_MODIFY_AUDIO_SETTINGS,
        K_READ_PHONE_STATE,
        K_WRITE_CONTACTS,
        K_READ_EXTERNAL_STORAGE,
        K_WRITE_EXTERNAL_STORAGE,
        K_CAMERA,
        K_READ_SMS,
        K_ACCESS_COARSE_LOCATION
    )
    annotation class Permission
    companion object {

        const val K_READ_CONTACTS = permission.READ_CONTACTS
        const val K_RECORD_AUDIO = permission.RECORD_AUDIO
        const val K_ACCESS_FINE_LOCATION = permission.ACCESS_FINE_LOCATION
        const val K_ACCESS_COARSE_LOCATION = permission.ACCESS_COARSE_LOCATION
        const val K_INTERNET = permission.INTERNET
        const val K_ACCESS_NETWORK_STATE = permission.ACCESS_NETWORK_STATE
        const val K_VIBRATE = permission.VIBRATE
        const val K_GET_ACCOUNTS = permission.GET_ACCOUNTS
        const val K_WAKE_LOCK = permission.WAKE_LOCK
        const val K_MODIFY_AUDIO_SETTINGS = permission.MODIFY_AUDIO_SETTINGS
        const val K_READ_PHONE_STATE = permission.READ_PHONE_STATE
        const val K_WRITE_CONTACTS = permission.WRITE_CONTACTS
        const val K_READ_EXTERNAL_STORAGE = permission.READ_EXTERNAL_STORAGE
        const val K_WRITE_EXTERNAL_STORAGE = permission.WRITE_EXTERNAL_STORAGE
        const val K_CAMERA = permission.CAMERA
        const val K_READ_SMS = permission.READ_SMS
        private val permissionRequests = ArrayList<PermissionRequest>()

        private fun checkPermission(context: Context, permission: String): Boolean {
            return VERSION.SDK_INT >= VERSION_CODES.M && ContextCompat.checkSelfPermission(context, permission) != PackageManager.PERMISSION_GRANTED
        }

        /**
         * Returns false if the Context does not have access to a all given permission.
         */
        @JvmStatic
        fun needPermission(context: Context, vararg permissions: String): Boolean {
            for (permission in permissions) {
                if (checkPermission(context, permission)) {
                    return true
                }
            }
            return false
        }

        /**
         * Returns true if the Context has access to a all given permission.
         */
        @JvmStatic
        fun hasPermission(context: Context, vararg permissions: String): Boolean {
            return !needPermission(context, *permissions)
        }

        @JvmStatic
        fun requestPermissions(
            activity: AppCompatCallbackActivity, permissions: Array<String>, callback: PermissionCallback = object : PermissionCallback {
                override fun permissionGranted() {}
                override fun permissionRefused() {}
            }
        ) {
            val permissionsHelper = PermissionsHelper()
            activity.addListener(permissionsHelper)
            permissionsHelper.askForPermission(activity, permissions, callback)
        }

        @JvmStatic
        fun requestPermission(
            activity: AppCompatCallbackActivity, permissions: String, callback: PermissionCallback = object : PermissionCallback {
                override fun permissionGranted() {}
                override fun permissionRefused() {}
            }
        ) {
            val permissionsHelper = PermissionsHelper()
            activity.addListener(permissionsHelper)
            permissionsHelper.askForPermission(activity, permissions, callback)
        }
    }
}